### ISSUE TITLE
Change 0018c884dbf228c54e30f0bc8b7586cc35c56b0f breaks conversion from time to money

### ISSUE THREAD

**Issue by Gumnos** [ledger/ledger#2031]
I used `git-bisect` to track down that [this commit causes breakage](https://github.com/ledger/ledger/commit/0018c884dbf228c54e30f0bc8b7586cc35c56b0f#commitcomment-50290475). Prior to this commit, I could convert time-log entries to dollars using my price-db. After this commit, it gives me errors. Sample/test files and the command-invocation are shown in that commit comment.

**Comment by Gumnos** [comment]
This was found [in this conversation over on Reddit](https://old.reddit.com/r/plaintextaccounting/comments/mzd6pd/converting_timelog_hours_to_usd_not_respecting/?context=10) and tracked down that it worked on Ledger version 3.1.3-20190331 which I had on OpenBSD 6.8, and broke when I upgraded packages from 6.8 to 6.9, bringing me to Ledger version 3.2.1-20200518. The latest version gives me different/updated error

    While parsing file "/home/gumnos/test.ledger", line 6:
    Error: Only one posting with null amount allowed per transaction

Which is misleading in a different direction.

**Comment by tbm** [comment]
Let's copy @daraul whose change broke this.  @daraul can you look into this please.

**Comment by tbm** [comment]
@daraul there's a test case: https://github.com/ledger/ledger/commit/0018c884dbf228c54e30f0bc8b7586cc35c56b0f#commitcomment-50290475

**Comment by daraul** [comment]
Oh great, thanks. I'm swamped atm so I can't promise it'll be fixed anytime soon, but it's on the TODO list.

**Comment by daraul** [comment]
I'm confused, but it looks as though this was fixed without me. @Gumnos, could you try reproducing the error on the master branch? I tried bisecting to find the “bad” commit that fixed it, but ended up at https://github.com/ledger/ledger/commit/051e4ab1ef82c58dc9563281ec14c02fc250e20f (which was made before you found this problem). I was able to confirm that 0018c884dbf228c54e30f0bc8b7586cc35c56b0f results in the error you mentioned, though.

**Comment by daraul** [comment]
I've written the test case out [here](https://github.com/daraul/ledger/commit/3631cab0cdcc19910964f8e31d7ab5ac5023924e), but I'd appreciate @Gumnos confirming whether or not `$37` is what he's expecting to see in `Assets:Bank:Checking`, based on his test command.

**Comment by daraul** [comment]
Sorry for the spam, but I must have mixed up my commands somewhere along the way. I just confirmed that the problem persists on master. I'll work on getting a fix together soon.

**Comment by Gumnos** [comment]
@daraul I was hoping to see $37.50 (or 37.49, depending on rounding). The initial `P` amount was determined by taking a $25/hr rate and converting it into a per-second rate (`25/(60*60)` ≈ $0.00694̅ per second, to within rounding tolerances). So I would expect the 1.5 hr to convert to 1.5*25 ≈ $37.50

But if limited to working in integer quantities, then yes, $37 is what I would otherwise expect to see when the resulting $37.49 rounds to the nearest dollar.  I'm not sure if there's a way to ensure that it's tracked down to pence (which would be preferable) rather than whole dollars.  Or even better, if I could specify my pricedb in hours

    P 2021-01-01 00:00:00 h 25

and then have the internal machinery convert dollars-per-hour to dollars-per-second for converting the timedb-internally-tracked-as-seconds to dollars.

**Comment by daraul** [comment]
Hmm, then I'm confused about the output of the test, then. It rounds to 37; maybe I should have a few more 0's [here](https://github.com/daraul/ledger/blob/3631cab0cdcc19910964f8e31d7ab5ac5023924e/test/regress/2031_a.dat#L1)? What's confusing is that it reports $37 in `Hours:Project` instead of `Assets:Bank:Checking`. I have no idea why that might be happening.

Test output: 
```
UpdateCTestConfiguration  from :/home/daraul/dev/ledger/DartConfiguration.tcl
UpdateCTestConfiguration  from :/home/daraul/dev/ledger/DartConfiguration.tcl
Test project /home/daraul/dev/ledger
Constructing a list of tests
Done constructing a list of tests
Updating test list for fixtures
Added 0 tests to meet fixture requirements
Checking test dependency graph...
Checking test dependency graph end
test 276
    Start 276: RegressTest_2031_a

276: Test command: /usr/bin/python3.8 "/home/daraul/dev/ledger/test/RegressTests.py" "/home/daraul/dev/ledger/ledger" "/home/daraul/dev/ledger" "/home/daraul/dev/ledger/test/regress/2031_a.test"
276: Environment variables: 
276:  TZ=America/Chicago
276: Test timeout computed to be: 10000000
276: FAILURE in output from /home/daraul/dev/ledger/test/regress/2031_a.test:
276: --
276: $ledger -f "/home/daraul/dev/ledger/test/regress/2031_a.test" --price-db test/regress/2031_a.dat -V bal
276: 
276: --
276:   @@ -1 +1 @@
276: 
276:   -                 $37  Assets:Bank:Checking
276: 
276:   +                 $37  Hours:Project
276: 
276: E[2031_a.test]STDERR:
276: b''
276: 
276: FAILED (1)
276: 
1/1 Test #276: RegressTest_2031_a ...............***Failed    0.04 sec

0% tests passed, 1 tests failed out of 1

Total Test time (real) =   0.05 sec

The following tests FAILED:
	276 - RegressTest_2031_a (Failed)
```

**Comment by Gumnos** [comment]
well, it would be additional "4"s, not "0"s (0.006944444444444) to get slightly more accurate, but even with just 2 "4"s worth, it should still come out to within pennies (the Python REPL tells me that `0.006944 * 1.5 * 60 * 60 = 37.4976` which is about the accuracy I'd expect from using floating-point math rather than the fixed decimal that I understand `ledger` uses).  So I suspect there's some internal truncation happening, even if it's just at formatting/display-time.  Do you get the pence if you specify

    commodity USD
        format $1,000.00

(where I presume the `format $1,000.00` specifies resolution to the penny)

**Comment by daraul** [comment]
Yes, I get the pence if I specify the commodity's formatting. I had no idea you could (or would) do this.

**Comment by Gumnos** [comment]
So my hope/expectation would be that

    commodity USD
        format $1,000.00

    P 2021-01-01 00:00:00 h 25 USD

    i 2021-04-29 07:00:00    Hours:Project
    o 2021-04-29 08:30:00

    2021-04-30 ! Get paid
      Hours:Project  = 0s
      Assets:Bank:Checking

when processed with

    $ ledger -f test.ledger bal

would report that `Hours:Project` has a 0 balance and `Assets:Bank:Checking` has $37.50.  That it's appearing in `Hours:Project` is just weird though.

I'm not sure what would be required to make that happen though.

**Comment by tbm** [comment]
@daraul @Gumnos what's the status of this?  Is this a bug/regression? Do things work as expected?

**Comment by daraul** [comment]
I would consider it a bug, and no, things don't yet work as expected. I saw you added it to 3.3, but I really can't promise I'll be able to get it fixed soon. From what I can tell, calculation of elided amounts may require changes.

**Comment by tbm** [comment]
@daraul I added it to 3.3 because it's a regression (right?) so ideally it would get fixed before the new release.

But the new release is overdue so I fear we may have to release without a fix

**Comment by daraul** [comment]
I understand. I appreciate both you guys being so patient.

**Comment by Gumnos** [comment]
no worries. I shifted to doing time-tracking outside of `ledger` (though I look forward to the time when I can shift back) so I can be indefinitely patient. :grin: 

Hopefully the test-case I provided should be sufficient to determine the pass/fail condition.

**Comment by dmage** [comment]
Unfortunately that's a flawed use case. There are two options how time can be transferred into the checking account.

The first option is to keep it as time. In this case, you'll see new values in your checking account every time market price for time changes. I doubt that's a desired result.

The second option is to explicitly convert time into dollars:

```
2021-04-30 ! Get paid
  Hours:Project  -5400s
  Assets:Bank:Checking  37.50 USD
```

But in this case you can write whatever you want. Ledger doesn't use pricedb to verify postings when they use different commodities. For example:

```
2021-04-30 ! Test
  Assets:Stocks       1 AAPL
  Assets:Checking   -50 USD
```

Ledger won't complain on this transaction regardless your pricedb.

The real bug is that timelog incorrectly interacts with normal transactions:

```
i 2021-04-29 07:00:00    Hours:Project
o 2021-04-29 08:30:00

2021-04-30 ! Test
  Hours:Project  -5400s = 0
  Hours:Paid      5400s
```

3.2.1 and master both say:

```
While parsing posting:
  Hours:Project  -5400s = 0
                          ^
Error: Balance assertion off by 5400s (expected to see -5400s)
```

**Comment by dmage** [comment]
`i` and `o` generate virtual postings (apparently because there is no counterpart account). With virtual accounts in the transaction it starts to work:

```
i 2021-04-29 07:00:00    Hours:Project
o 2021-04-29 08:30:00

2021-04-30 ! Test
  [Hours:Project]  = 0
  [Hours:Paid]
```

So the minimal reproducer is

```
2021-04-29
  (Foo)  10 USD

2021-04-30
  Foo  -10 USD = 0
  Bar  10 USD
```

I doubt if it is a valid ledger though.

**Comment by tbm** [comment]
Removing milestone since it's unlikely this will be fixed soon.

**Comment by daraul** [comment]
@tbm / @Gumnos apologies for the terrible response time -- I found myself with some time, and was able to work out https://github.com/ledger/ledger/pull/2336, but it's got some problems. Looking forward to any feedback on the solution.
